---
title: "hbc_scRNA-seq training - integration"
output: html_document
---

# Single-cell RNA-seq clustering analysis: Integration

```{r setup, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 12, fig.height = 7)
```

### Learning Objectives:

-   Perform integration of cells across conditions to identify cells that are similar to each other
-   Describe complex integration tasks and alternative tools for integration

***Goals:***

-   *To **align same cell types** across conditions.*

***Challenges:***

-   ***Aligning cells of similar cell types** so that we do not have clustering downstream due to differences between samples, conditions, modalities, or batches*

***Recommendations:***

-   *Go through the analysis without integration first to determine whether integration is necessary*

------------------------------------------------------------------------

## To integrate or not no integrate?

Clustering without integration first to decide whether or not to perform alignment. For this workshop, in the normalization step, normalization were performed on both conditions together in a Seurat object `seurat_phase`, we can create an UMAP to see if there is a sample-specific clustering:

```{r run UMAP}
# For this object we have already run PCA
# Run UMAP
seurat_phase <- RunUMAP(seurat_phase,
dims = 1:40,reduction = "pca")
# Plot UMAP
DimPlot(seurat_phase)

```

Condition-specific clustering of the cells indicates that we need to integrate the cells across conditions to ensure that cells of the same type cluster together.

This will enable more interpretable results downstream(i.e. DE analysis, ligand-receptor analysis, differential abundance analysis ...)

### Integrate or align samples across conditions using shared highly variable genes

### Integrating using CCA(canonical correlation analysis) in the Seurat package

*Steps applied:* 1. Perform canonical correlation analysis CCA is a form of PCA, but it identifies the greatest sources of variation in the data and only kept if it is shared or conserved across the conditions/samples(using the 3000 most variant genes from each sample)

2.  Identify anchors or mutual nearest neighbors(MNN) across datasets (sometimes incorrect anchors are identified)

3.  Filter anchors

4.  Integrate the conditions

Use anchors and corresponding scores to transform the cell expression values, allowing for the integration of the conditions/datasets (different samples, conditions, datasets, modalities)

> NOTE: Transformation of each cell uses a weighted average of the two cells of each anchor across anchors of the datasets. Weights determined by cell similarity score (distance between cell and k nearest anchors) and anchor scores, so cells in the same neighborhood should have similar correction values.

## Perform the integration across conditions
1. Select the most variable features to use for integration
```{r}
# specify using all of the 3000 most variable genes identified by SCTransform by set nfeatures
integ_features <- SelectIntegrationFeatures(object.list = split_seurat, 
                                            nfeatures = 3000) 
```

2. Prepare the SCTransform object for integration.
```{r}
split_seurat <- PrepSCTIntegration(object.list = split_seurat, 
                                   anchor.features = integ_features)
```
3. Perform CCA, find the best buddies or anchors and filter incorrect anchors.

```{r perfrom CCA}
# Find best buddies - can take a while to run
integ_anchors <- FindIntegrationAnchors(object.list = split_seurat, 
                                        normalization.method = "SCT", 
                                        anchor.features = integ_features)
```
4. Integrate across conditions

```{r}
# Integrate across conditions
seurat_integrated <- IntegrateData(anchorset = integ_anchors, 
                                   normalization.method = "SCT")
```

### UMAP visualization after integration
```{r visualize with PCA}
# Run PCA
seurat_integrated <- RunPCA(object = seurat_integrated)

# Plot PCA
PCAPlot(seurat_integrated,
        split.by = "sample")  

```

```{r visualize with UMAP}

# Set seed to avoid the creation of a slightly different umap each time
# usually set seed at the beginning of the script
set.seed(123456)

# Run UMAP
seurat_integrated <- RunUMAP(seurat_integrated, 
                             dims = 1:40,
			     reduction = "pca")

# Plot UMAP                             
DimPlot(seurat_integrated)                             
```

*Side-by-side comparison of clusters*
Split the plotting between conditions for easier comparison
```{r}
# Plot UMAP split by sample
DimPlot(seurat_integrated,
        split.by = "sample")  
```


```{r}
# Save integrated seurat object
saveRDS(seurat_integrated, "results/integrated_seurat.rds")
```
